addons/pos_pinelabs/models/pos_payment_method.py

```diff
@@ -0,0 +1,117 @@
+
+    def _get_payment_terminal_selection(self):
+        return super()._get_payment_terminal_selection() + [('pinelabs', 'Pine Labs')]
+
+    def pinelabs_make_payment_request(self, data):
+        pinelabs = PinelabsPosRequest(self)
+        body = pinelabs._pinelabs_request_body()
+        body.update({
+            'Amount': data.get('amount'),
+            'TransactionNumber': data.get('transactionNumber'),
+            'SequenceNumber': data.get('sequenceNumber')
+        })
+        response = pinelabs._call_pinelabs(endpoint='UploadBilledTransaction', payload=body)
+        if not response.get('ResponseCode') and response.get('ResponseMessage') == "APPROVED":
+            return {
+                'responseCode': response.get('ResponseCode'),
+                'status': response.get('ResponseMessage'),
+                'plutusTransactionReferenceID': response.get('PlutusTransactionReferenceID'),
+            }
+        default_error_msg = _('The expected error code for the Pine Labs POS payment upload request was not included in the response.')
+        error = response.get('ResponseMessage') or default_error_msg
+        return {"error": error}
+
+    def pinelabs_fetch_payment_status(self, data):
+        pinelabs = PinelabsPosRequest(self)
+        body = pinelabs._pinelabs_request_body(payment_mode=False)
+        body.update({'PlutusTransactionReferenceID': data.get('plutusTransactionReferenceID')})
+
+        retry_count = 0
+        def get_status():
+            nonlocal retry_count
+            with self.pool.cursor() as cr:
+                new_env = self.env(cr=cr)
+                config = new_env["pos.config"].browse(data.get("config_id"))
+                payment_method = new_env["pos.payment.method"].browse(data.get("payment_method_id"))
+                try:
+                    response = pinelabs._call_pinelabs(endpoint='GetCloudBasedTxnStatus', payload=body)
+                    if response.get('ResponseCode') == 1001:
+                        if retry_count <= MAX_RETRIES:
+                            retry_count += 1
+                            Timer(5, get_status).start()
+                        else:
+                            # We automatically cancel transactions in cases of inactivity.
+                            error = _("The transaction has failed due to inactivity")
+                            payment_method.pinelabs_cancel_payment_request(data)
+                            config._notify("PINELABS_PAYMENT_RESPONSE", {"error": error})
+                    elif response.get('ResponseCode') == 0:
+                        pinelabs_latest_response = {
+                            'responseCode': response.get('ResponseCode'),
+                            'status': response.get('ResponseMessage'),
+                            'plutusTransactionReferenceID': response.get('PlutusTransactionReferenceID'),
+                            'data': format_transaction_data(response.get('TransactionData'))
```

Comment:
if there is any sensitive information here we should not send it via the bus. You can send a simple message on the bus to notify the client that the payment succeeded for ex and if the client needs the rest of the data you can do an rpc call to get it.